File: app/src/main/java/com/example/asthmaapplication/main/homepage/LoginFragment.java
package com.example.asthmaapplication.main.homepage;

import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;
import androidx.lifecycle.ViewModelProvider;

import com.example.asthmaapplication.R;
import com.example.asthmaapplication.databinding.FragmentLoginBinding;
import com.example.asthmaapplication.main.common.BaseFragment;
import com.example.asthmaapplication.main.mainpage.MainActivity;
import com.example.asthmaapplication.main.utils.UIUtils;


import javax.annotation.Nullable;
import javax.inject.Inject;

import dagger.hilt.android.AndroidEntryPoint;
import io.reactivex.annotations.NonNull;

@AndroidEntryPoint
public class LoginFragment extends BaseFragment {
    FragmentLoginBinding binding;
    HomePageViewModel viewModel;
    FragmentManager manager;
    FragmentTransaction transaction;
    TextWatcher watcher;

    @Inject
    public LoginFragment() {

    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        viewModel = new ViewModelProvider(this).get(HomePageViewModel.class);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentLoginBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@Nullable View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        getPreferences();
        setActionBarTitle();

        binding.actionLogin.setEnabled(false);
        UIUtils.addUnderlineFlag(binding.actionRegister);

        manager = getFragmentManager();
        transaction = manager.beginTransaction();

        binding.actionRegister.setOnClickListener(v -> {
            launchRegistrationPage();
        });

        watcher = new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable s) {
                boolean allowLogin = (!TextUtils.isEmpty(binding.emailLogin.getText().toString()) &&
                        !TextUtils.isEmpty(binding.passwordLogin.getText().toString()) &&
                        !TextUtils.isEmpty(binding.nameLogin.getText().toString()));
                binding.actionLogin.setEnabled(allowLogin);
                binding.actionLogin.setOnClickListener(v -> {
                    viewModel.login(binding.emailLogin.getText().toString(),
                            binding.passwordLogin.getText().toString());
                    setUserNamePreferences(binding.nameLogin.getText().toString());
                    resetLoginPage();
                    launchUserLoginSuccessful();
                });
            }
        };

        binding.emailLogin.addTextChangedListener(watcher);
        binding.passwordLogin.addTextChangedListener(watcher);
        binding.nameLogin.addTextChangedListener(watcher);

    }

    @Override
    public void onResume() {
        super.onResume();
    }

    public void resetLoginPage() {
        binding.emailLogin.setText(null);
        binding.passwordLogin.setText(null);
        binding.nameLogin.setText(null);
    }

    public void launchUserLoginSuccessful() {
        Intent intent = new Intent(getContext(), MainActivity.class);
        startActivity(intent);
    }

    public void launchRegistrationPage() {
        RegistrationFragment fragment = new RegistrationFragment();
        transaction
                .addToBackStack(null)
                .replace(R.id.fragment_container, fragment)
                .commit();
    }

    public void setActionBarTitle() {
        setActionBarTitle(getString(R.string.login_page_title));
    }

    @Override
    public View getRoot() {
        return binding.getRoot();
    }
}

File: app/src/main/java/com/example/asthmaapplication/main/homepage/LoginFragment.kt
package com.example.asthmaapplication.main.homepage

import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.viewModels
import com.example.asthmaapplication.R
import com.example.asthmaapplication.databinding.FragmentLoginBinding
import com.example.asthmaapplication.main.common.BaseActivity
import com.example.asthmaapplication.main.common.BaseFragment
import com.example.asthmaapplication.main.common.Constants
import com.example.asthmaapplication.main.common.Event
import com.example.asthmaapplication.main.mainpage.MainActivity
import com.example.asthmaapplication.main.utils.ActivityUtils
import com.example.asthmaapplication.main.utils.UIUtils
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
class LoginFragment: BaseFragment() {
    private lateinit var _binding: FragmentLoginBinding
    private lateinit var watcher: TextWatcher

    private val _viewModel: HomePageViewModel by viewModels()
    private var _currentUser: String? = null
    private var _userType: String? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        _binding = FragmentLoginBinding.inflate(inflater, container, false)
        return _binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setActionBarTitle()
        setupUi()

        _viewModel.snackBar.observe(viewLifecycleOwner) { showSnackBar(it) }
        _viewModel.loginSuccess.observe(viewLifecycleOwner) { launchMainPage(it) }
    }

    override fun onResume() {
        super.onResume()
        _currentUser = _viewModel.getCurrentUser()
        setupListeners()
    }

    private fun setupUi() {
        (activity as BaseActivity?)?.showBackButton(true)
        UIUtils.addUnderlineFlag(_binding.actionRegister)
        UIUtils.addUnderlineFlag(_binding.actionContinue)
    }

    private fun setupListeners() {
        _binding.actionLogin.isEnabled = false
        _binding.actionRegister.setOnClickListener { launchRegistrationPage() }

        if (!_currentUser.isNullOrEmpty()) {
            _binding.actionContinue.setOnClickListener { launchMainPage() }
        } else {
            _binding.actionContinue.text = getString(R.string.continue_as_guest)
            _binding.actionContinue.setOnClickListener {
                _viewModel.clearCurrentUser()
                launchMainPage()
            }
        }

        _binding.loginOptionGroup.setOnCheckedChangeListener{ _, checkedType ->
            _userType = when (checkedType) {
                R.id.student_login -> {
                    Constants.STUDENT_OPTION
                }
                R.id.patient_login -> {
                    Constants.PATIENT_OPTION
                }
                else -> {
                    null
                }
            }
        }

        watcher = object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(s: Editable) {
                val allowLogin = !TextUtils.isEmpty(_binding.emailLogin.text.toString()) &&
                        !TextUtils.isEmpty(_binding.passwordLogin.text.toString()) &&
                        !TextUtils.isEmpty(_binding.nameLogin.text.toString())
                _binding.actionLogin.isEnabled = allowLogin

                _binding.actionLogin.setOnClickListener {
                    _viewModel.login(
                        _binding.emailLogin.text.toString(),
                        _binding.passwordLogin.text.toString(),
                        _binding.nameLogin.text.toString(),
                        _userType
                    )
                }
            }
        }
        _binding.emailLogin.addTextChangedListener(watcher)
        _binding.passwordLogin.addTextChangedListener(watcher)
        _binding.nameLogin.addTextChangedListener(watcher)
    }

    private fun resetLoginPage() {
        _binding.emailLogin.text = null
        _binding.passwordLogin.text = null
        _binding.nameLogin.text = null
    }

    private fun launchMainPage(event: Event<Boolean>) {
        event.contentIfNotHandled?.let {
            if (it) {
                launchMainPage()
                resetLoginPage()
            }
        }
    }

    private fun launchMainPage() {
        val intent = Intent(context, MainActivity::class.java)
        startActivity(intent)
    }

    private fun launchRegistrationPage() {
        val fragment = RegistrationFragment()
        ActivityUtils.addFragmentWithBackStack(
            parentFragmentManager,
            fragment,
            R.id.fragment_container,
            null
        )
    }

    private fun setActionBarTitle() {
        setActionBarTitle(getString(R.string.login_page_title))
    }

    override fun getRoot(): View {
        return _binding.root
    }
}